src/message/ping-parser.h

```diff
@@ -30,84 +44,139 @@ class PingParser
+        if (data == 'R') {
+            rxBuffer_[rxCount_++] = data;
             state_++;
-            break;
-        case WAIT_LENGTH_H:
-            rxBuffer_[rxCount_++] = b;
-            payloadLength_ = static_cast<uint16_t>((b << 8) | payloadLength_);
-            if (payloadLength_ <= rxBufferLength_ - 8 - 2) {
-                state_++;
-            } else {
-                state_ = WAIT_START;
-            }
-            break;
-        case WAIT_MSG_ID_L: // fall-through
-        case WAIT_MSG_ID_H:
-        case WAIT_SRC_ID:
-        case WAIT_DST_ID:
-            rxBuffer_[rxCount_++] = b;
+        } else {
+            reset();
+        }
+        break;
+    case PingParser::State::WAIT_LENGTH_L:
+        rxBuffer_[rxCount_++] = data;
+        payloadLength_ = data;
+        state_++;
+        break;
+    case PingParser::State::WAIT_LENGTH_H:
+        rxBuffer_[rxCount_++] = data;
+        payloadLength_ = static_cast<uint16_t>((data << 8) | payloadLength_);
+        if (payloadLength_ <= rxBufferLength_ - 8 - 2) {
+            state_++;
+        } else {
+            reset();
+        }
+        break;
+    case PingParser::State::WAIT_MSG_ID_L: // fall-through
+    case PingParser::State::WAIT_MSG_ID_H:
+    case PingParser::State::WAIT_SRC_ID:
+    case PingParser::State::WAIT_DST_ID:
+        rxBuffer_[rxCount_++] = data;
+        state_++;
+        if (payloadLength_ == 0) {
+            // no payload bytes, so we skip WAIT_PAYLOAD state
             state_++;
-            if (payloadLength_ == 0) {
-                // no payload bytes, so we skip WAIT_PAYLOAD state
-                state_++;
-            }
-            break;
-        case WAIT_PAYLOAD:
-            rxBuffer_[rxCount_++] = b;
-            if (--payloadLength_ == 0) {
-                state_++;
-            }
-            break;
-        case WAIT_CHECKSUM_L:
-            rxBuffer_[rxCount_++] = b;
+        }
+        break;
+    case PingParser::State::WAIT_PAYLOAD:
+        rxBuffer_[rxCount_++] = data;
+        if (--payloadLength_ == 0) {
             state_++;
-            break;
-        case WAIT_CHECKSUM_H:
-            rxBuffer_[rxCount_++] = b;
-            state_ = WAIT_START;
-            if (rxMessage.verifyChecksum()) {
-                parsed++;
-                return NEW_MESSAGE;
-            } else {
-                errors++;
-                return ERROR;
-            }
         }
-        return (ParseState)state_;
+        break;
+    case PingParser::State::WAIT_CHECKSUM_L:
+        rxBuffer_[rxCount_++] = data;
+        state_++;
+        break;
+    case PingParser::State::WAIT_CHECKSUM_H:
+        rxBuffer_[rxCount_++] = data;
+        state_ = PingParser::State::WAIT_START;
+        if (rxMessage.verifyChecksum()) {
+            parsed++;
+            return PingParser::State::NEW_MESSAGE;
+        } else {
+            errors++;
+            return PingParser::State::ERROR;
+        }
+    default:
```

Comment:
I'd remove the return state_; in the default as for me it decreases readability, because there's a return state_; just after the switch. On a quick look one migth think "hum, the state_ returnerd in the switch and outside are the same, why there's two returns?", but the switch sometimes modify the satte, sometimes not.

I'd remove the default implementation to make that explicit. (and perhaps change the implementation of the switch to contain function calls instead of code so it's easier to reason about it.

Reply:
The idea of this patch is to increase the readability a bit more and not refactor the code, I believe that we can accomplish this in future PRs.